grr/server/grr_response_server/flows/general/osquery.py

```diff
@@ -67,13 +103,55 @@ def _UpdateProgress(
     self.state.progress.partial_table = _GetTruncatedTable(responses)
     self.state.progress.total_row_count = _GetTotalRowCount(responses)
 
+  def _GetPathSpecsToCollect(
+      self,
+      responses: flow_responses.Responses[rdf_osquery.OsqueryResult],
+  ) -> List[rdf_paths.PathSpec]:
+    columns_for_collection = self.args.file_collect_columns
+    if len(columns_for_collection) == 0:
+      return []
+
+    if len(columns_for_collection) > FILE_COLLECT_LIMIT_MAX_COLUMNS:
+      message = (f"Requested file collection for {len(columns_for_collection)} "
+          f"columns, but the limit is {FILE_COLLECT_LIMIT_MAX_COLUMNS} columns")
+      raise _FileCollectionLimitsExceeded(message)
+
+    total_row_count = _GetTotalRowCount(responses)
+    if total_row_count > FILE_COLLECT_LIMIT_MAX_ROWS:
+      message = (f"Requested file collection on a table with {total_row_count} "
+          f"rows, but the limit is {FILE_COLLECT_LIMIT_MAX_ROWS} rows")
+      raise _FileCollectionLimitsExceeded(message)
+
+    file_names = []
+    for osquery_result in responses:
+      for column in columns_for_collection:
+        file_names.extend(osquery_result.table.Column(column))
+
+    return list(map(_PathSpecFromFileName, file_names))
+
+  def _FileCollectionFromColumns(
+      self,
+      responses: flow_responses.Responses[rdf_osquery.OsqueryResult],
+  ) -> None:
+    pathspecs = self._GetPathSpecsToCollect(responses)
+    self.CallFlow(
```

Comment:
How would we limit the total number of bytes allowed? As you suggested, I used `network_bytes_limit` as a parameter to `CallClient`, but I'm not sure where this would fit here since we are not calling a separate flow?

Also how can we speed it up by using `MultiGetLogic`? I don't know too much about their implementation, and to me it sounds like we are going to implement new `MultiGetFile` ourselves if we use `MultiGetLogic`?

Reply:
> I don't know too much about their implementation, and to me it sounds like we are going to implement new MultiGetFile ourselves if we use MultiGetLogic?

MultiGetFile is indeed a thin layer on top of the MultiGetFileLogic and, unfortunately, not the best one. The biggest issue - MultiGetFile doesn't report errors correctly. Another issue: there's no way you can influence MultiGetFile's truncation logic. With MultiGetFileLogic you could, for every file, first do a StatFile client action call to check its side and then make a decision whether you want to download it.

By "suboptimal" I meant that MultiGetFile has known issues and, coincidentally, using MultiGetFileLogic offers a workaround around these issues. Fundamentally, I think we need to fix MultiGetFile to make it usable in such scenarios as we have here. I don't think we should do any changes within this PR.

Reply:
Using `network_bytes_limit` will, IIUC, have no effect on MultiGetFile. It also has a "file_size" parameter: https://github.com/google/grr/blob/252af5d85f02121a75f8d83ebbee8e3d7e4f5fc9/grr/proto/grr_response_proto/flows.proto#L1348

But it will silently truncate the files instead of skipping.